std/string.d

```diff
@@ -2937,26 +2937,96 @@ if (isForwardRange!Range && isSomeChar!(ElementEncodingType!Range) &&
+        // with a single wchar and due to the design of UTF-16 those wchars
+        // will not occur as part of the encoding of multi-wchar codepoints.
+        static if (isDynamicArray!Range || (isRandomAccessRange!Range && hasLength!Range
+            && __traits(compiles, { size_t s = input.length / 2; input = input[s .. $]; })))
         {
-            if (!std.ascii.isWhite(c))
-                break;
-            input.popFront();
+            size_t i = 0;
+            for (const size_t end = input.length; i < end; ++i)
+            {
+                if (!std.uni.isWhite(input[i]))
+                    break;
+            }
+            input = input[i .. $];
+            return input;
         }
         else
         {
-            auto save = input.save;
-            auto dc = decodeFront(input);
-            if (!std.uni.isWhite(dc))
-                return save;
+            while (!input.empty)
+            {
+                if (!std.uni.isWhite(input.front))
+                    break;
+                input.popFront();
+            }
+            return input;
+        }
+    }
+    else
+    {
+        static if (isDynamicArray!Range || (isRandomAccessRange!Range && hasLength!Range
+            && __traits(compiles, { size_t s = input.length / 2; input = input[s .. $]; })))
+        {{
+            // ASCII optimization for dynamic arrays & similar.
+            size_t i = 0;
+            for (const size_t end = input.length; i < end; ++i)
+            {
+                auto c = input[i];
+                if (c >= 0x80) goto NonAsciiPath;
+                if (!std.ascii.isWhite(c)) break;
+            }
+            input = input[i .. $];
+            return input;
+
+        NonAsciiPath:
+            input = input[i .. $];
+            // Fall through to standard case.
+        }}
+
+        static if (ElementType!Range.sizeof > 1)
+        {
+            // Type performs its own decoding.
+            while (!input.empty)
+            {
+                if (!std.uni.isWhite(input.front))
+                    break;
+                input.popFront();
+            }
+            return input;
+        }
+        else
+        {
+            // Type doesn't perform its own decoding. We impose decoding on it
+            // which is questionable but is how this function has worked up to
+            // now.
+            import std.utf : decodeFront, UseReplacementDchar;
+            while (!input.empty)
+            {
+                auto c = input.front;
+                if (std.ascii.isASCII(c))
+                {
+                    if (!std.ascii.isWhite(c))
+                        break;
+                    input.popFront();
+                }
+                else
+                {
+                    auto save = input.save;
+                    auto dc = decodeFront!(UseReplacementDchar.yes)(input);
+                    if (!std.uni.isWhite(dc))
+                        return save;
```

Comment:
This section of code isn't tested by the given unittests.

Reply:
Added unittest.